src/core/egg-herbie.rkt

```diff
@@ -570,21 +617,25 @@
     (define type->id (hash-ref! egg-id->table eid (lambda () (make-hasheq))))
     (hash-ref! type->id type (lambda () (new-eclass eid type))))
 
-  ;; build typed eclasses
-  (for ([eclass (in-list egraph)])
-    (define eid (car eclass))
-    (for ([enode (in-list (cdr eclass))])
-      (let ([enode (normalize-enode enode egg->herbie)])
-        ; get all possible types for the enode
-        (define types (enode-type enode egg->herbie))
-        (for ([type (in-list types)])
-          ; lookup or create a new typed eclass
-          (define id (lookup-id eid type))
-          (define enode* (rebuild-enode enode type lookup-id))
-          (define eclass (hash-ref id->eclass id))
-          (set-box! eclass (cons enode* (unbox eclass)))))))
-
-  (define eclasses (list->vector (map (compose reverse unbox) (reverse eclass-boxes))))
+  ;; extract (untyped) eclass ids as u32vector
+  ;; for each eclass, extract the enodes
+  ;;  <enode> ::= <symbol>
+  ;;            | <number>
+  ;;            | (<symbol> . <u32vector>)
+  ;; NOTE: nodes in typed eclasses are reversed relative
+  ;; to their position in untyped eclasses
+  (for ([eid (in-u32vector (egraph-eclasses egraph-data))])
+    (for ([enode (in-vector (egraph-get-eclass egraph-data eid))])
+      ; get all possible types for the enode
+      ; lookup its correct eclass and add the rebuilt node
+      (define types (enode-type enode egg->herbie))
+      (for ([type (in-list types)])
+        (define id (lookup-id eid type))
+        (define enode* (rebuild-enode enode type lookup-id))
+        (define eclass (hash-ref id->eclass id))
```

Comment:
Drive-by comment by why is this a hash table instead of a vector? Hash tables in Racket are slow.

Reply:
I keep trying to removing hash tables, but there are two problems: (i) Rust-side (untyped) eclass ids are not sequential, and (ii) don't know how many typed eclass ids we need unless we consume the whole data structure. I've been thinking about this a bit recently, and I think I'll take one last stab at it.

Reply:
Oh, I see, `lookup-id` returns a compound structure? My bad. I thought it was just a number. (For "how many", you can always just make a _really big_ vector? NumEclass * NumReprs?